File: leetcode/javascript/backtracking/palindromePartitioning(medium)/palindromePartitioning.js
/* https://leetcode.com/problems/palindrome-partitioning/description/

Дана строка s, раздел s такой, что каждый подстрока раздела представляет собой палиндром.
Вернуть все возможные палиндромные разбиения s.

Input: s = "aab"
Output: [["a","a","b"],["aa","b"]]

Input: s = "a"
Output: [["a"]]
*/

/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function (s) {
    const result = [];

    const backtrack = (start, path) => {
        if (start === s.length) {
            result.push([...path]);
            return;
        }

        for (let end = start; end < s.length; end++) {
            if (isPalindrome(s, start, end)) {
                path.push(s.slice(start, end + 1));
                backtrack(end + 1, path);
                path.pop();
            }
        }
    };

    const isPalindrome = (str, left, right) => {
        while (left < right) {
            if (str[left++] !== str[right--]) return false;
        }
        return true;
    };

    backtrack(0, []);
    return result;
};

console.log(partition("aab")); // [["a","a","b"],["aa","b"]]

File: leetcode/go/backtracking/palindromePartitioning(medium)/palindromePartitioning.go
package main

import "fmt"

/* https://leetcode.com/problems/palindrome-partitioning/description/

Дана строка s, раздел s такой, что каждый подстрока раздела представляет собой палиндром.
Вернуть все возможные палиндромные разбиения s.

Input: s = "aab"
Output: [["a","a","b"],["aa","b"]]

Input: s = "a"
Output: [["a"]]
*/

func main() {
	fmt.Println(partition("aab")) // [["a","a","b"],["aa","b"]]
}

// partition возвращает все возможные разбиения строки на палиндромы
func partition(s string) [][]string {
	var result [][]string

	// Внутренняя функция для проверки палиндрома
	isPalindrome := func(str string, left, right int) bool {
		for left < right {
			if str[left] != str[right] {
				return false
			}
			left++
			right--
		}
		return true
	}

	// Рекурсивная функция для генерации разбиений
	var backtrack func(int, []string)
	backtrack = func(start int, path []string) {
		// Если дошли до конца строки, добавляем текущее разбиение в результат
		if start == len(s) {
			// Создаем копию пути, чтобы избежать изменений
			temp := make([]string, len(path))
			copy(temp, path)
			result = append(result, temp)
			return
		}

		// Перебираем все возможные подстроки, начиная с текущей позиции
		for end := start; end < len(s); end++ {
			// Если подстрока - палиндром, продолжаем рекурсивно
			if isPalindrome(s, start, end) {
				// Добавляем палиндром в текущий путь
				path = append(path, s[start:end+1])
				// Рекурсивно обрабатываем оставшуюся часть строки
				backtrack(end+1, path)
				// Удаляем последний элемент (backtracking)
				path = path[:len(path)-1]
			}
		}
	}

	backtrack(0, []string{})
	return result
}
